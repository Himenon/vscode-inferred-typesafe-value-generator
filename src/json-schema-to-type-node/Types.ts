import type { JSONSchema7, JSONSchema7TypeName } from "json-schema";

export type Schema = JSONSchema7;

export type JSONSchemaTypeName = JSONSchema7TypeName;

export type JSONSchemaDefinition = Schema | boolean;

export interface Reference {
  $ref: string;
  summary?: string;
  description?: string;
}

export interface UnSupportSchema extends Omit<Schema, "type"> {
  type: JSONSchemaTypeName[];
}

export interface OneOfSchema extends Omit<Schema, "oneOf"> {
  oneOf: Schema[];
}

export interface AllOfSchema extends Omit<Schema, "allOf"> {
  allOf: Schema[];
}

export interface AnyOfSchema extends Omit<Schema, "anyOf"> {
  anyOf: Schema[];
}

export interface MultiTypeSchema extends Omit<Schema, "type"> {
  type: JSONSchemaTypeName[];
}

export interface ObjectSchema extends Omit<Schema, "type"> {
  type: "object";
}

export interface FunctionSchema extends Omit<ObjectSchema, "format"> {
  format: "function";
}

export interface ObjectSchemaWithAdditionalProperties extends ObjectSchema {
  additionalProperties: JSONSchemaDefinition;
}

export interface ArraySchema extends Omit<Schema, "type"> {
  type: "array";
}

export interface PrimitiveSchema extends Omit<Schema, "type"> {
  type: "string" | "number" | "integer" | "boolean" | "null";
}

export interface AnySchema extends Omit<Schema, "type"> {
  type: "any";
}
