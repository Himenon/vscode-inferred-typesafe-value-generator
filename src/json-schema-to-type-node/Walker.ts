import * as TypeGuard from "./TypeGuard";
import type * as Types from "./Types";
import * as DotProp from "dot-prop";
import * as Logger from "../logger";

export interface ConvertedObject {
  value: any;
  errorMessages: string[];
}

export class Walker {
  private readonly schema: Types.Schema;
  constructor(schema: Types.Schema) {
    this.schema = JSON.parse(JSON.stringify(schema));
  }

  private _convert = (schema?: Types.Schema): any => {
    if (!schema) {
      return "undefined";
    }
    if (TypeGuard.isReference(schema)) {
      return this.reference(schema);
    }
    if (TypeGuard.isObjectSchema(schema)) {
      if (TypeGuard.isFunctionSchema(schema)) {
        return this.functionTypeValue(schema);
      }
      return this.convertObject(schema);
    }
    if (TypeGuard.isPrimitiveSchema(schema)) {
      return this.primitive(schema);
    }
    if (TypeGuard.isArraySchema(schema)) {
      return this.array(schema);
    }
    if (TypeGuard.isMultiTypeSchema(schema)) {
      return this.multiTypeValue(schema);
    }
    if (TypeGuard.isAnyOfSchema(schema)) {
      return this.anyOfValue(schema);
    }
    if (TypeGuard.isOneOfSchema(schema)) {
      return this.oneOfValue(schema);
    }
    if (TypeGuard.isAllOfSchema(schema)) {
      return this.allOfValue(schema);
    }
    throw new Error(`UnSupport JSON Schema: ${JSON.stringify(schema)}`);
  };

  public convert = (schema?: Types.Schema): ConvertedObject => {
    try {
      const value = this._convert(schema);
      return {
        value,
        errorMessages: [],
      };
    } catch (error) {
      const e = error as Error;
      Logger.error(e.stack);
      return {
        value: undefined,
        errorMessages: [e.message],
      };
    }
  };

  private undefinedValue = (): any => {
    return undefined;
  };

  private reference = (schema: Types.Reference): any => {
    const decodeRef = decodeURIComponent(schema.$ref);
    const ref = decodeRef.replace(/\./g, "\\.").replace(/^#\//, "").replace(/\//g, ".");
    const refSchema = DotProp.get(this.schema, ref) as any;
    return this._convert(refSchema);
  };

  private array = (schema: Types.ArraySchema): any[] => {
    const items = schema.items;
    if (!items || items === true) {
      return [];
    }
    if (!Array.isArray(items)) {
      return [this._convert(items)];
    }
    return items.map((item) => {
      if (typeof item === "boolean") {
        return "";
      }
      return this._convert(item);
    });
  };

  private allOfValue = (schema: Types.AllOfSchema) => {
    if (schema.allOf.length > 0) {
      const allOfObject = schema.allOf.reduce((previous, current) => {
        const nextSchema = this._convert(current);
        return { ...previous, ...nextSchema };
      }, {});
      return allOfObject;
    }
    return this.undefinedValue();
  };

  private oneOfValue = (schema: Types.OneOfSchema) => {
    if (schema.oneOf.length > 0) {
      return this._convert(schema.oneOf[0]);
    }
    return this.undefinedValue();
  };

  private anyOfValue = (schema: Types.AnyOfSchema) => {
    if (schema.anyOf.length > 0) {
      return this._convert(schema.anyOf[0]);
    }
    return this.undefinedValue();
  };

  private multiTypeValue = (schema: Types.MultiTypeSchema) => {
    const types = schema.type;
    if (types.length > 0) {
      return this._convert({ type: types[0] });
    }
    return this.undefinedValue();
  };

  private stringValue = (schema: Types.PrimitiveSchema): string => {
    let estimateValue: any;
    if (schema.const) {
      estimateValue = schema.const;
    } else if (schema.default) {
      estimateValue = schema.default;
    } else if (schema.enum && schema.enum.length > 0) {
      estimateValue = schema.enum[0];
    }
    if (typeof estimateValue === "string") {
      return estimateValue;
    }
    return "";
  };

  private numberValue = (schema: Types.PrimitiveSchema): number => {
    let estimateValue: any;
    if (schema.const) {
      estimateValue = schema.const;
    } else if (schema.default) {
      estimateValue = schema.default;
    } else if (schema.enum && schema.enum.length > 0) {
      estimateValue = schema.enum[0];
    }
    if (estimateValue) {
      if (typeof estimateValue === "string") {
        const value = parseInt(estimateValue, 10);
        return value || 0;
      } else if (typeof estimateValue === "number") {
        return estimateValue;
      }
    }
    return 0;
  };

  private booleanValue = (schema: Types.PrimitiveSchema): boolean => {
    let estimateValue: any;
    if (schema.default) {
      estimateValue = schema.default;
    } else if (schema.enum && schema.enum.length > 0) {
      estimateValue = schema.enum[0];
    }
    if (typeof estimateValue === "boolean") {
      return estimateValue;
    }
    if (typeof estimateValue === "string") {
      return Boolean(estimateValue);
    }
    return false;
  };

  private primitive = (schema: Types.PrimitiveSchema) => {
    switch (schema.type) {
      case "boolean":
        return this.booleanValue(schema);
      case "integer":
      case "number":
        return this.numberValue(schema);
      case "string":
        return this.stringValue(schema);
      default:
        throw new Error("UNSupport Type!");
    }
  };

  private convertObject = (schema: Types.ObjectSchema) => {
    const properties = schema.properties || {};
    const obj: Record<string, any> = {};
    Object.entries(properties).forEach(([key, value]) => {
      if (typeof value === "boolean") {
        return;
      }
      obj[key] = this._convert(value);
    });
    return obj;
  };

  /**
   * TODO
   */
  private functionTypeValue = (_schema: Types.FunctionSchema) => {
    return undefined;
  };
}
