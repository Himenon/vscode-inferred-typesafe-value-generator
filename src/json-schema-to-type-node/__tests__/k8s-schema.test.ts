import * as Module from "../Walker";
import { Schema } from "../Types";
import schema from "./k8s-schema.json";

const jsonSchema = schema as Schema;

describe("Kubernetes Schema", () => {
  test("Kubernetes Schema", () => {
    const walker = new Module.Walker(jsonSchema);
    const { value } = walker.convert(jsonSchema);
    expect(value).toMatchSnapshot();
  });
});
