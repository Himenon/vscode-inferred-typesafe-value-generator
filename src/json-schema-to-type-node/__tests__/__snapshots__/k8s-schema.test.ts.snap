// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kubernetes Schema Kubernetes Schema 1`] = `
Object {
  "apiVersion": "",
  "kind": "",
  "metadata": Object {
    "annotations": Object {},
    "clusterName": "",
    "creationTimestamp": "",
    "deletionGracePeriodSeconds": 0,
    "deletionTimestamp": "",
    "finalizers": Array [],
    "generateName": "",
    "generation": 0,
    "labels": Object {},
    "managedFields": Array [
      Object {
        "apiVersion": "",
        "fieldsType": "",
        "fieldsV1": Object {},
        "manager": "",
        "operation": "",
        "subresource": "",
        "time": undefined,
      },
    ],
    "name": "",
    "namespace": "",
    "ownerReferences": Array [
      Object {
        "apiVersion": "",
        "blockOwnerDeletion": false,
        "controller": false,
        "kind": "",
        "name": "",
        "uid": "",
      },
    ],
    "resourceVersion": "",
    "selfLink": "",
    "uid": "",
  },
  "spec": Object {
    "minReadySeconds": 0,
    "paused": false,
    "progressDeadlineSeconds": 0,
    "replicas": 0,
    "revisionHistoryLimit": 0,
    "selector": Object {
      "matchExpressions": Array [
        Object {
          "key": "",
          "operator": "",
          "values": Array [],
        },
      ],
      "matchLabels": Object {},
    },
    "strategy": Object {
      "rollingUpdate": Object {
        "maxSurge": 0,
        "maxUnavailable": 0,
      },
      "type": "",
    },
    "template": Object {
      "metadata": Object {
        "annotations": Object {},
        "clusterName": "",
        "creationTimestamp": undefined,
        "deletionGracePeriodSeconds": 0,
        "deletionTimestamp": undefined,
        "finalizers": Array [],
        "generateName": "",
        "generation": 0,
        "labels": Object {},
        "managedFields": Array [
          Object {
            "apiVersion": "",
            "fieldsType": "",
            "fieldsV1": Object {},
            "manager": "",
            "operation": "",
            "subresource": "",
            "time": undefined,
          },
        ],
        "name": "",
        "namespace": "",
        "ownerReferences": Array [
          Object {
            "apiVersion": "",
            "blockOwnerDeletion": false,
            "controller": false,
            "kind": "",
            "name": "",
            "uid": "",
          },
        ],
        "resourceVersion": "",
        "selfLink": "",
        "uid": "",
      },
      "spec": Object {
        "activeDeadlineSeconds": 0,
        "affinity": Object {
          "nodeAffinity": Object {
            "preferredDuringSchedulingIgnoredDuringExecution": Array [
              Object {
                "preference": Object {
                  "matchExpressions": Array [
                    Object {
                      "key": "",
                      "operator": "",
                      "values": Array [],
                    },
                  ],
                  "matchFields": Array [
                    Object {
                      "key": "",
                      "operator": "",
                      "values": Array [],
                    },
                  ],
                },
                "weight": 0,
              },
            ],
            "requiredDuringSchedulingIgnoredDuringExecution": Object {
              "nodeSelectorTerms": Array [
                Object {
                  "matchExpressions": Array [],
                  "matchFields": Array [],
                },
              ],
            },
          },
          "podAffinity": Object {
            "preferredDuringSchedulingIgnoredDuringExecution": Array [
              Object {
                "podAffinityTerm": Object {
                  "labelSelector": Object {
                    "matchExpressions": Array [
                      Object {
                        "key": "",
                        "operator": "",
                        "values": Array [],
                      },
                    ],
                    "matchLabels": Object {},
                  },
                  "namespaceSelector": undefined,
                  "namespaces": Array [],
                  "topologyKey": "",
                },
                "weight": 0,
              },
            ],
            "requiredDuringSchedulingIgnoredDuringExecution": Array [
              Object {
                "labelSelector": undefined,
                "namespaceSelector": undefined,
                "namespaces": Array [],
                "topologyKey": "",
              },
            ],
          },
          "podAntiAffinity": Object {
            "preferredDuringSchedulingIgnoredDuringExecution": Array [
              Object {
                "podAffinityTerm": undefined,
                "weight": 0,
              },
            ],
            "requiredDuringSchedulingIgnoredDuringExecution": Array [],
          },
        },
        "automountServiceAccountToken": false,
        "containers": Array [
          Object {
            "args": Array [],
            "command": Array [],
            "env": Array [
              Object {
                "name": "",
                "value": "",
                "valueFrom": Object {
                  "configMapKeyRef": Object {
                    "key": "",
                    "name": "",
                    "optional": false,
                  },
                  "fieldRef": Object {
                    "apiVersion": "",
                    "fieldPath": "",
                  },
                  "resourceFieldRef": Object {
                    "containerName": "",
                    "divisor": "",
                    "resource": "",
                  },
                  "secretKeyRef": Object {
                    "key": "",
                    "name": "",
                    "optional": false,
                  },
                },
              },
            ],
            "envFrom": Array [
              Object {
                "configMapRef": Object {
                  "name": "",
                  "optional": false,
                },
                "prefix": "",
                "secretRef": Object {
                  "name": "",
                  "optional": false,
                },
              },
            ],
            "image": "",
            "imagePullPolicy": "",
            "lifecycle": Object {
              "postStart": Object {
                "exec": Object {
                  "command": Array [],
                },
                "httpGet": Object {
                  "host": "",
                  "httpHeaders": Array [
                    Object {
                      "name": "",
                      "value": "",
                    },
                  ],
                  "path": "",
                  "port": undefined,
                  "scheme": "",
                },
                "tcpSocket": Object {
                  "host": "",
                  "port": undefined,
                },
              },
              "preStop": Object {
                "exec": Object {
                  "command": Array [],
                },
                "httpGet": Object {
                  "host": "",
                  "httpHeaders": Array [
                    Object {
                      "name": "",
                      "value": "",
                    },
                  ],
                  "path": "",
                  "port": undefined,
                  "scheme": "",
                },
                "tcpSocket": Object {
                  "host": "",
                  "port": undefined,
                },
              },
            },
            "livenessProbe": Object {
              "exec": undefined,
              "failureThreshold": 0,
              "httpGet": undefined,
              "initialDelaySeconds": 0,
              "periodSeconds": 0,
              "successThreshold": 0,
              "tcpSocket": undefined,
              "terminationGracePeriodSeconds": 0,
              "timeoutSeconds": 0,
            },
            "name": "",
            "ports": Array [
              Object {
                "containerPort": 0,
                "hostIP": "",
                "hostPort": 0,
                "name": "",
                "protocol": "",
              },
            ],
            "readinessProbe": Object {
              "exec": undefined,
              "failureThreshold": 0,
              "httpGet": undefined,
              "initialDelaySeconds": 0,
              "periodSeconds": 0,
              "successThreshold": 0,
              "tcpSocket": undefined,
              "terminationGracePeriodSeconds": 0,
              "timeoutSeconds": 0,
            },
            "resources": Object {
              "limits": Object {},
              "requests": Object {},
            },
            "securityContext": Object {
              "allowPrivilegeEscalation": false,
              "capabilities": Object {
                "add": Array [],
                "drop": Array [],
              },
              "privileged": false,
              "procMount": "",
              "readOnlyRootFilesystem": false,
              "runAsGroup": 0,
              "runAsNonRoot": false,
              "runAsUser": 0,
              "seLinuxOptions": Object {
                "level": "",
                "role": "",
                "type": "",
                "user": "",
              },
              "seccompProfile": Object {
                "localhostProfile": "",
                "type": "",
              },
              "windowsOptions": Object {
                "gmsaCredentialSpec": "",
                "gmsaCredentialSpecName": "",
                "hostProcess": false,
                "runAsUserName": "",
              },
            },
            "startupProbe": undefined,
            "stdin": false,
            "stdinOnce": false,
            "terminationMessagePath": "",
            "terminationMessagePolicy": "",
            "tty": false,
            "volumeDevices": Array [
              Object {
                "devicePath": "",
                "name": "",
              },
            ],
            "volumeMounts": Array [
              Object {
                "mountPath": "",
                "mountPropagation": "",
                "name": "",
                "readOnly": false,
                "subPath": "",
                "subPathExpr": "",
              },
            ],
            "workingDir": "",
          },
        ],
        "dnsConfig": Object {
          "nameservers": Array [],
          "options": Array [
            Object {
              "name": "",
              "value": "",
            },
          ],
          "searches": Array [],
        },
        "dnsPolicy": "",
        "enableServiceLinks": false,
        "ephemeralContainers": Array [
          Object {
            "args": Array [],
            "command": Array [],
            "env": Array [
              Object {
                "name": "",
                "value": "",
                "valueFrom": Object {
                  "configMapKeyRef": Object {
                    "key": "",
                    "name": "",
                    "optional": false,
                  },
                  "fieldRef": Object {
                    "apiVersion": "",
                    "fieldPath": "",
                  },
                  "resourceFieldRef": Object {
                    "containerName": "",
                    "divisor": "",
                    "resource": "",
                  },
                  "secretKeyRef": Object {
                    "key": "",
                    "name": "",
                    "optional": false,
                  },
                },
              },
            ],
            "envFrom": Array [
              Object {
                "configMapRef": Object {
                  "name": "",
                  "optional": false,
                },
                "prefix": "",
                "secretRef": Object {
                  "name": "",
                  "optional": false,
                },
              },
            ],
            "image": "",
            "imagePullPolicy": "",
            "lifecycle": Object {
              "postStart": undefined,
              "preStop": undefined,
            },
            "livenessProbe": undefined,
            "name": "",
            "ports": Array [
              Object {
                "containerPort": 0,
                "hostIP": "",
                "hostPort": 0,
                "name": "",
                "protocol": "",
              },
            ],
            "readinessProbe": undefined,
            "resources": Object {
              "limits": Object {},
              "requests": Object {},
            },
            "securityContext": Object {
              "allowPrivilegeEscalation": false,
              "capabilities": Object {
                "add": Array [],
                "drop": Array [],
              },
              "privileged": false,
              "procMount": "",
              "readOnlyRootFilesystem": false,
              "runAsGroup": 0,
              "runAsNonRoot": false,
              "runAsUser": 0,
              "seLinuxOptions": Object {
                "level": "",
                "role": "",
                "type": "",
                "user": "",
              },
              "seccompProfile": Object {
                "localhostProfile": "",
                "type": "",
              },
              "windowsOptions": Object {
                "gmsaCredentialSpec": "",
                "gmsaCredentialSpecName": "",
                "hostProcess": false,
                "runAsUserName": "",
              },
            },
            "startupProbe": undefined,
            "stdin": false,
            "stdinOnce": false,
            "targetContainerName": "",
            "terminationMessagePath": "",
            "terminationMessagePolicy": "",
            "tty": false,
            "volumeDevices": Array [
              Object {
                "devicePath": "",
                "name": "",
              },
            ],
            "volumeMounts": Array [
              Object {
                "mountPath": "",
                "mountPropagation": "",
                "name": "",
                "readOnly": false,
                "subPath": "",
                "subPathExpr": "",
              },
            ],
            "workingDir": "",
          },
        ],
        "hostAliases": Array [
          Object {
            "hostnames": Array [],
            "ip": "",
          },
        ],
        "hostIPC": false,
        "hostNetwork": false,
        "hostPID": false,
        "hostname": "",
        "imagePullSecrets": Array [
          Object {
            "name": "",
          },
        ],
        "initContainers": Array [
          Object {
            "args": Array [],
            "command": Array [],
            "env": Array [],
            "envFrom": Array [],
            "image": "",
            "imagePullPolicy": "",
            "lifecycle": undefined,
            "livenessProbe": undefined,
            "name": "",
            "ports": Array [],
            "readinessProbe": undefined,
            "resources": undefined,
            "securityContext": undefined,
            "startupProbe": undefined,
            "stdin": false,
            "stdinOnce": false,
            "terminationMessagePath": "",
            "terminationMessagePolicy": "",
            "tty": false,
            "volumeDevices": Array [],
            "volumeMounts": Array [],
            "workingDir": "",
          },
        ],
        "nodeName": "",
        "nodeSelector": Object {},
        "overhead": Object {},
        "preemptionPolicy": "",
        "priority": 0,
        "priorityClassName": "",
        "readinessGates": Array [
          Object {
            "conditionType": "",
          },
        ],
        "restartPolicy": "",
        "runtimeClassName": "",
        "schedulerName": "",
        "securityContext": Object {
          "fsGroup": 0,
          "fsGroupChangePolicy": "",
          "runAsGroup": 0,
          "runAsNonRoot": false,
          "runAsUser": 0,
          "seLinuxOptions": undefined,
          "seccompProfile": undefined,
          "supplementalGroups": Array [],
          "sysctls": Array [
            Object {
              "name": "",
              "value": "",
            },
          ],
          "windowsOptions": undefined,
        },
        "serviceAccount": "",
        "serviceAccountName": "",
        "setHostnameAsFQDN": false,
        "shareProcessNamespace": false,
        "subdomain": "",
        "terminationGracePeriodSeconds": 0,
        "tolerations": Array [
          Object {
            "effect": "",
            "key": "",
            "operator": "",
            "tolerationSeconds": 0,
            "value": "",
          },
        ],
        "topologySpreadConstraints": Array [
          Object {
            "labelSelector": undefined,
            "maxSkew": 0,
            "topologyKey": "",
            "whenUnsatisfiable": "",
          },
        ],
        "volumes": Array [
          Object {
            "awsElasticBlockStore": Object {
              "fsType": "",
              "partition": 0,
              "readOnly": false,
              "volumeID": "",
            },
            "azureDisk": Object {
              "cachingMode": "",
              "diskName": "",
              "diskURI": "",
              "fsType": "",
              "kind": "",
              "readOnly": false,
            },
            "azureFile": Object {
              "readOnly": false,
              "secretName": "",
              "shareName": "",
            },
            "cephfs": Object {
              "monitors": Array [],
              "path": "",
              "readOnly": false,
              "secretFile": "",
              "secretRef": Object {
                "name": "",
              },
              "user": "",
            },
            "cinder": Object {
              "fsType": "",
              "readOnly": false,
              "secretRef": undefined,
              "volumeID": "",
            },
            "configMap": Object {
              "defaultMode": 0,
              "items": Array [
                Object {
                  "key": "",
                  "mode": 0,
                  "path": "",
                },
              ],
              "name": "",
              "optional": false,
            },
            "csi": Object {
              "driver": "",
              "fsType": "",
              "nodePublishSecretRef": undefined,
              "readOnly": false,
              "volumeAttributes": Object {},
            },
            "downwardAPI": Object {
              "defaultMode": 0,
              "items": Array [
                Object {
                  "fieldRef": undefined,
                  "mode": 0,
                  "path": "",
                  "resourceFieldRef": undefined,
                },
              ],
            },
            "emptyDir": Object {
              "medium": "",
              "sizeLimit": undefined,
            },
            "ephemeral": Object {
              "volumeClaimTemplate": Object {
                "metadata": undefined,
                "spec": Object {
                  "accessModes": Array [],
                  "dataSource": Object {
                    "apiGroup": "",
                    "kind": "",
                    "name": "",
                  },
                  "dataSourceRef": Object {
                    "apiGroup": "",
                    "kind": "",
                    "name": "",
                  },
                  "resources": undefined,
                  "selector": undefined,
                  "storageClassName": "",
                  "volumeMode": "",
                  "volumeName": "",
                },
              },
            },
            "fc": Object {
              "fsType": "",
              "lun": 0,
              "readOnly": false,
              "targetWWNs": Array [],
              "wwids": Array [],
            },
            "flexVolume": Object {
              "driver": "",
              "fsType": "",
              "options": Object {},
              "readOnly": false,
              "secretRef": undefined,
            },
            "flocker": Object {
              "datasetName": "",
              "datasetUUID": "",
            },
            "gcePersistentDisk": Object {
              "fsType": "",
              "partition": 0,
              "pdName": "",
              "readOnly": false,
            },
            "gitRepo": Object {
              "directory": "",
              "repository": "",
              "revision": "",
            },
            "glusterfs": Object {
              "endpoints": "",
              "path": "",
              "readOnly": false,
            },
            "hostPath": Object {
              "path": "",
              "type": "",
            },
            "iscsi": Object {
              "chapAuthDiscovery": false,
              "chapAuthSession": false,
              "fsType": "",
              "initiatorName": "",
              "iqn": "",
              "iscsiInterface": "",
              "lun": 0,
              "portals": Array [],
              "readOnly": false,
              "secretRef": undefined,
              "targetPortal": "",
            },
            "name": "",
            "nfs": Object {
              "path": "",
              "readOnly": false,
              "server": "",
            },
            "persistentVolumeClaim": Object {
              "claimName": "",
              "readOnly": false,
            },
            "photonPersistentDisk": Object {
              "fsType": "",
              "pdID": "",
            },
            "portworxVolume": Object {
              "fsType": "",
              "readOnly": false,
              "volumeID": "",
            },
            "projected": Object {
              "defaultMode": 0,
              "sources": Array [
                Object {
                  "configMap": Object {
                    "items": Array [
                      Object {
                        "key": "",
                        "mode": 0,
                        "path": "",
                      },
                    ],
                    "name": "",
                    "optional": false,
                  },
                  "downwardAPI": Object {
                    "items": Array [
                      Object {
                        "fieldRef": undefined,
                        "mode": 0,
                        "path": "",
                        "resourceFieldRef": undefined,
                      },
                    ],
                  },
                  "secret": Object {
                    "items": Array [],
                    "name": "",
                    "optional": false,
                  },
                  "serviceAccountToken": Object {
                    "audience": "",
                    "expirationSeconds": 0,
                    "path": "",
                  },
                },
              ],
            },
            "quobyte": Object {
              "group": "",
              "readOnly": false,
              "registry": "",
              "tenant": "",
              "user": "",
              "volume": "",
            },
            "rbd": Object {
              "fsType": "",
              "image": "",
              "keyring": "",
              "monitors": Array [],
              "pool": "",
              "readOnly": false,
              "secretRef": undefined,
              "user": "",
            },
            "scaleIO": Object {
              "fsType": "",
              "gateway": "",
              "protectionDomain": "",
              "readOnly": false,
              "secretRef": undefined,
              "sslEnabled": false,
              "storageMode": "",
              "storagePool": "",
              "system": "",
              "volumeName": "",
            },
            "secret": Object {
              "defaultMode": 0,
              "items": Array [],
              "optional": false,
              "secretName": "",
            },
            "storageos": Object {
              "fsType": "",
              "readOnly": false,
              "secretRef": undefined,
              "volumeName": "",
              "volumeNamespace": "",
            },
            "vsphereVolume": Object {
              "fsType": "",
              "storagePolicyID": "",
              "storagePolicyName": "",
              "volumePath": "",
            },
          },
        ],
      },
    },
  },
  "status": Object {
    "availableReplicas": 0,
    "collisionCount": 0,
    "conditions": Array [
      Object {
        "lastTransitionTime": undefined,
        "lastUpdateTime": undefined,
        "message": "",
        "reason": "",
        "status": "",
        "type": "",
      },
    ],
    "observedGeneration": 0,
    "readyReplicas": 0,
    "replicas": 0,
    "unavailableReplicas": 0,
    "updatedReplicas": 0,
  },
}
`;
